helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/WagedInstanceCapacity.java

```diff
@@ -0,0 +1,164 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+import javax.annotation.Nullable;
+
+import org.apache.helix.controller.rebalancer.util.WagedValidationUtil;
+import org.apache.helix.controller.stages.CurrentStateOutput;
+import org.apache.helix.model.ClusterConfig;
+import org.apache.helix.model.InstanceConfig;
+import org.apache.helix.model.Message;
+import org.apache.helix.model.Partition;
+import org.apache.helix.model.ResourceConfig;
+import org.apache.helix.controller.dataproviders.InstanceCapacityDataProvider;
+import org.apache.helix.controller.dataproviders.ResourceControllerDataProvider;
+import org.apache.helix.controller.rebalancer.util.WagedValidationUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class WagedInstanceCapacity implements InstanceCapacityDataProvider {
+  private static final Logger LOG = LoggerFactory.getLogger(WagedInstanceCapacity.class);
+
+  // Available Capacity per Instance
+  private Map<String, Map<String, Integer>> _perInstance;
+  private ResourceControllerDataProvider _cache;
+
+  public WagedInstanceCapacity(ResourceControllerDataProvider clusterData) {
+    _cache = clusterData;
+    _perInstance = new HashMap<>();
+
+    ClusterConfig clusterConfig = _cache.getClusterConfig();
+    for (InstanceConfig instanceConfig : _cache.getInstanceConfigMap().values()) {
+      Map<String, Integer> instanceCapacity =
+        WagedValidationUtil.validateAndGetInstanceCapacity(clusterConfig, instanceConfig);
+      _perInstance.put(instanceConfig.getInstanceName(), instanceCapacity);
+    }
+  }
+
+  /**
+   * Create Default Capacity Map.
+   * This is a utility method to create a default capacity map matching instance capacity map for participants.
+   * This is required as non-WAGED partitions will be placed on same instance and we don't know their actual capacity.
+   * This will generate default values of 0 for all the capacity keys.
+   */
+  public Map<String, Integer> createDefaultParticipantWeight() {
+    // copy the value of first Instance capacity.
+    Map<String, Integer> partCapacity = new HashMap<>(_perInstance.values().iterator().next());
+
+    // reset the value of all capacity to 0.
+    for (String key : partCapacity.keySet()) {
+      partCapacity.put(key, 0);
+    }
+    return partCapacity;
+  }
+
+  /**
+   * Process the pending messages based on the Current states
+   * @param currentState - Current state of the resources.
+   */
+  public void processPendingMessages(CurrentStateOutput currentState) {
```

Comment:
As the top comment goes, this feels a little deviate from the responsibility of this class.

Reply:
+1 to this.
Everything else makes sense to me but this method looks like one method which goes over "all" resources and then reduce available instance capacity. 
First we can update method name that suggests so. (may be like iteratePendingMsgsAndUpdateInstanceCapacity so something like that). And if that's too long name then atleast description should be updated.